wandb/sdk/launch/sweeps/scheduler_optuna.py

```diff
@@ -0,0 +1,297 @@
+                )
+            else:
+                print(f"{LOG_PREFIX} Unknown parameter type, help! {param=}, {extras=}")
+        return config, trial
+
+    def _optuna_heartbeat(self, agent_id: str) -> List[Dict[str, Any]]:
+        def objective(trial):
+            """Optuna objective function"""
+            sweep_config, trial = self.make_trial(trial)
+            queued_run: QueuedRun = self._add_to_launch_queue(
+                entry_point=[
+                    f"WANDB_SWEEP_ID={self._sweep_id} python3",
+                    self._sweep_config.get("program"),
+                ],
+                # Use legacy sweep utilities to extract args dict from agent heartbeat run.args (?)
+                config={
+                    "overrides": {
+                        "run_config": LegacySweepAgent._create_command_args(
+                            {"args": sweep_config}
+                        )["args_dict"],
+                    },
+                    "resource": "local-process",
+                },
+            )
+
+            # wait for launch to actually create a run object
+            launched_run = queued_run.wait_until_running()
+            launched_run_path = "/".join(launched_run.path)
+
+            encoded_run_id = base64.standard_b64encode(
+                f"Run:v1:{launched_run.id}:{self._project}:{self._entity}".encode()
+            ).decode("utf-8")
+
+            api_run = self._public_api.run(launched_run_path)
+            # Set external runs sweep information
+            self._internal_api.upsert_run(id=encoded_run_id, sweep_name=self._sweep_id)
+
+            last_epoch = -1
+            while True:
+                history = api_run.scan_history(keys=["_step", "loss_metric"])
+                next_epochs = [x for x in history if x["_step"] > last_epoch]
+
+                if len(next_epochs) == 0:
+                    time.sleep(1)
+                    continue
+
+                for epoch_data in next_epochs:
+                    step = epoch_data["_step"]
+                    loss = epoch_data["loss_metric"]
+                    sweep_job.log({"step": step, "loss": loss})
+                    print(f"{LOG_PREFIX}Step {step}: loss={loss}")
+                    trial.report(loss, step)  # todo fix metric
+
+                last_epoch = step
+                if trial.should_prune():
+                    print(f"{LOG_PREFIX}Optuna decided to PRUNE!")
+                    success = self._internal_api._stop_run(encoded_run_id)
+                    if success:
+                        print(f"{LOG_PREFIX}Stopped run: {api_run.id}")
+                    else:
+                        print(f"{LOG_PREFIX}Couldn't stop run...")
+
+                    raise optuna.TrialPruned()
+
+                if last_epoch == trial.params["epochs"] - 1:
+                    print(f"{LOG_PREFIX} Trial completed!")
+                    sweep_job.log({"loss": loss})
+                    return loss
```

Comment:
All of this would have to be in ADDITION to whatever the user has in their own objective() function. We would also need to scrub mentions of wandb like wandb.init or wandb.log... @KyleGoyette